File: 2025/ruby/day02.rb
require_relative 'day'

class Day02 < Day # >

  # @example
  #   day.part1 #=> 1227775554
  def part1
    ranges = parse_input
    invalid_ids = []

    ranges.each do |range|
      range.each do |id|
        invalid_ids << id if repeats_twice?(id)
      end
    end

    invalid_ids
      .reduce(&:+)
  end

  # @example
  #   day.part2 #=> 4174379265
  def part2
    ranges = parse_input
    invalid_ids = []

    ranges.each do |range|
      range.each do |id|
        invalid_ids << id if repeats_at_least_twice?(id)
      end
    end

    invalid_ids
      .reduce(&:+)
  end


  def repeats_twice?(id)
    id_str = id.to_s
    return false if id_str.length.odd?
    return true if id_str[0...(id_str.length/2)] == id_str[(id_str.length/2)..]
    false
  end

  # @example 11
  #   day.repeats_at_least_twice? 11 #=> true
  # @example 22
  #   day.repeats_at_least_twice? 22 #=> true
  # @example 99
  #   day.repeats_at_least_twice? 99 #=> true
  # @example 111
  #   day.repeats_at_least_twice? 111 #=> true
  # @example 999
  #   day.repeats_at_least_twice? 999 #=> true
  # @example 1010
  #   day.repeats_at_least_twice? 1010 #=> true
  # @example 565656
  #   day.repeats_at_least_twice? 565656 #=> true
  # @example 824824824
  #   day.repeats_at_least_twice? 824824824 #=> true
  def repeats_at_least_twice?(id)
    id_str = id.to_s
    (id_str + id_str)[1..-2].include?(id_str)
  end

  # @example
  #   day = Day02.new(EXAMPLE_INPUT)
  #   parsed = day.parse_input
  #   parsed.first #=> (11..22)
  def parse_input
    @input
      .split(",")
      .map { |range|
        Range.new(*(range.split("-").map(&:to_i)))
      }
  end

  EXAMPLE_INPUT = File.read("../inputs/day02-example-input.txt")
end

File: 2025/go/day02/day02.go
package day02

import (
	"fmt"
	"log"
	"robbkidd/aoc2025/utils"
	"strconv"
	"strings"
)

func Run() {
	input, err := utils.ReadInput()
	if err != nil {
		log.Fatal("Couldn't read real input")
	}
	fmt.Printf("Part1: %s\n", Part1(input))
	fmt.Printf("Part2: %s\n", Part2(input))
}

func Part1(input string) string {
	ranges := parse(input)
	invalid_id_sum := 0
	for _, r := range ranges {
		for id := r.start; id <= r.end; id++ {
			if repeats_twice(id) {
				invalid_id_sum += id
			}
		}
	}
	return strconv.Itoa(invalid_id_sum)
}

func repeats_twice(id int) bool {
	id_str := strconv.Itoa(id)
	if len(id_str)%2 > 0 {
		return false
	}
	return id_str[:len(id_str)/2] == id_str[len(id_str)/2:]
}

func Part2(input string) string {
	ranges := parse(input)
	invalid_id_sum := 0
	for _, r := range ranges {
		for id := r.start; id <= r.end; id++ {
			if repeats_at_least_twice(id) {
				invalid_id_sum += id
			}
		}
	}
	return strconv.Itoa(invalid_id_sum)
}

func repeats_at_least_twice(id int) bool {
	id_str := strconv.Itoa(id)
	doubled := id_str + id_str
	doubled_and_snipped := doubled[1 : len(doubled)-1]
	return strings.Contains(doubled_and_snipped, id_str)
}

type Range struct {
	start, end int
}

func parse(input string) []Range {
	ranges_in := strings.Split(input, ",")
	ranges := make([]Range, len(ranges_in))

	for i, range_in := range ranges_in {
		bounds := strings.Split(range_in, "-")
		r := Range{}
		r.start, _ = strconv.Atoi(bounds[0])
		r.end, _ = strconv.Atoi(bounds[1])
		ranges[i] = r
	}
	return ranges
}
